lsp/nls/src/requests/completion.rs

```diff
@@ -29,18 +29,129 @@ use crate::{
 // We follow the path by traversing a term, type or contract which represents a record
 // and stop when there is nothing else on the path
 
+// Type or contract
+pub enum TorC {
+    T(Types),
+    C(RichTerm),
+}
+
+// #[derive(Eq, PartialEq, Debug, Ord, PartialOrd, Default)]
+struct IdentWithMeta {
+    ident: Ident,
+    info: TorC,
+    item: Option<LinearizationItem<Types>>,
+}
+
+impl From<Ident> for IdentWithMeta {
+    fn from(ident: Ident) -> Self {
+        IdentWithMeta {
+            ident,
+            info: TorC::T(Types(TypeF::Dyn)),
+            item: None,
+        }
+    }
+}
+
+impl From<&str> for IdentWithMeta {
+    fn from(ident: &str) -> Self {
+        IdentWithMeta {
+            ident: Ident::from(ident),
+            info: TorC::T(Types(TypeF::Dyn)),
+            item: None,
+        }
+    }
+}
+
+impl IdentWithMeta {
+    fn compute_detail(&self) -> String {
+        let info = || match &self.info {
+            TorC::T(ty) => ty.to_string(),
+            TorC::C(term) => term.to_string(),
+        };
```

Comment:
Maybe `TOrC` should implement `to_string()`? But once again, if you just use `Types`, this part isn't even needed.